interfaces/builtin/docker.go

```diff
@@ -0,0 +1,185 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"github.com/snapcore/snapd/interfaces"
+)
+
+var dockerPermanentSlotAppArmor = []byte(`
+# Description: Allow operating as the Docker daemon. Reserved because this
+#  gives privileged access to the system.
+# Usage: reserved
+
+# Allow sockets
+/{,var/}run/docker.sock rw,
+/{,var/}run/docker/     rw,
+/{,var/}run/docker/**   mrwklix,
+/{,var/}run/runc/       rw,
+/{,var/}run/runc/**     mrwklix,
+
+# Wide read access to /proc, but somewhat limited writes for now
+@{PROC}/ r,
+@{PROC}/** r,
+@{PROC}/[0-9]*/attr/exec w,
+@{PROC}/sys/net/** w,
+@{PROC}/[0-9]*/cmdline r,
+@{PROC}/[0-9]*/oom_score_adj w,
+
+# Wide read access to /sys
+/sys/** r,
+# Limit cgroup writes a bit
+/sys/fs/cgroup/*/docker/   rw,
+/sys/fs/cgroup/*/docker/** rw,
+/sys/fs/cgroup/*/system.slice/   rw,
+/sys/fs/cgroup/*/system.slice/** rw,
+
+# Allow tracing other processes (especially ourselves)
+ptrace (read, trace),
+
+# Docker needs a lot of caps, but limits them in the app container
+capability,
+
+# Allow talking to systemd
+dbus (send)
+    bus=system
+    peer=(name=org.freedesktop.systemd*,label=unconfined),
+# Allow receiving from unconfined
+dbus (receive)
+    bus=system
+    peer=(label=unconfined),
+
+# Docker does all kinds of mounts all over the filesystem
+/dev/mapper/control rw,
+/dev/mapper/docker* rw,
+/dev/loop* r,
+/dev/loop[0-9]* w,
+mount,
+umount,
+pivot_root,
+/.pivot_root*/ rw,
+
+# for console access
+/dev/ptmx rw,
+# file descriptors (/proc/NNN/fd/X)
+/[0-9]*   rw,
+
+# Docker needs to be able to create and load the profile it applies to containers ("docker-default")
+# XXX We might be able to get rid of this if we generate and load docker-default ourselves and make docker not do it.
+/etc/apparmor.d/docker rw,
+/sbin/apparmor_parser ixr,
+/etc/apparmor*/** r,
+/etc/apparmor.d/cache/docker* w,
+/sys/kernel/security/apparmor/** rw,
+
+# We'll want to adjust this to support --security-opts...
+change_profile -> docker-default,
+signal (send) peer=docker-default,
+ptrace (read, trace) peer=docker-default,
+
+# Graph (storage) driver bits
+/dev/shm/aufs.xino rw,
+
+#cf bug 1502785
+/ r,
+`)
+
+var dockerConnectedPlugAppArmor = []byte(`
+# Description: Allow using Docker. Reserved because this gives
+#  privileged access to the service/host.
+# Usage: reserved
+
+# Obviously need to be able to talk to the daemon
+/{,var/}run/docker.sock rw,
+
+@{PROC}/sys/net/core/somaxconn r,
+`)
```

Comment:
Right, I mean that the daemon will run as root, it will see that the "docker" group exists, and will apply it to the unix socket created.  Then, from the client side, only clients that have _both_ the "docker" group and the appropriate AppArmor permission will be able to talk to the Docker socket (and thus control the daemon).  If the "docker" group does not exist, the daemon will instead use "root", and so from the perspective of the users (or even other snaps) only root will be able to even connect to the Docker daemon.

IMO, anything that isn't `docker` itself (and perhaps a manually-reviewed set of related utilities) _should_ require manual connection of that `docker` plug, especially given the system-wide implications.